esteid/mobileid/base.py

```diff
@@ -0,0 +1,423 @@
+        assert hash_type in HASH_ALGORITHMS
+
+        message = str(message or self.msg("display_text"))
+        assert len(message) <= 20, f"Display text can not exceed 20 chars: got '{message}'"
+
+        if language not in Languages.ALL:
+            language = Languages.ENG
+
+        content_hash = generate_hash(hash_type, signed_data)
+        hash_value_b64 = base64.b64encode(content_hash)
+
+        try:
+            result = self.invoke(
+                self.Actions.SIGN,
+                method="POST",
+                data=self.rp_params(
+                    {
+                        "nationalIdentityNumber": id_code,
+                        "phoneNumber": phone_number,
+                        "hashType": hash_type,
+                        "hash": hash_value_b64.decode("utf-8"),
+                        "language": language,
+                        # Casting to str to ensure translations are resolved
+                        "displayText": str(message or self.msg("display_text")),  # NOTE: 20-char limit
+                        "displayTextFormat": "UCS-2",  # the other choice is GSM-7 which is 7-bit
+                    }
+                ),
+            )
+
+        except requests.HTTPError:
+            raise
+
+        return SignResult(
+            session_id=result["sessionID"],
+            digest=content_hash,
+            verification_code=get_verification_code(content_hash),
+        )
+
+    def sign_status(self, session_id, certificate: bytes, signed_digest, timeout=10000):
+        """Retrieve signing session result from Mobile-ID backend
+
+        The certificate is missing in the response, so
+        to verify the signature you must pass it on from the authentication process.
+
+        see https://github.com/SK-EID/MID#33-status-of-signing-and-authentication
+
+        :param session_id: session ID from I{sign} Result
+        :param certificate: the certificate used to sign data
+        :param signed_digest: the digest of data to sign
+        :param int timeout: Request long poll timeout value in milliseconds (Note: server uses a default
+                         if client does not send it)
+        :rtype: SignStatusResult
+        """
+        data = self._get_session_response(self.Actions.SIGN, session_id, timeout)
+
+        signature = base64.b64decode(data["signature"]["value"])
+        signature_algorithm = data["signature"]["algorithm"]
+        assert signature_algorithm[:6].upper() in HASH_ALGORITHMS
+
+        try:
+            pyasice.verify(certificate, signature, signed_digest, signature_algorithm[:6], prehashed=True)
+        except pyasice.SignatureVerificationError:
+            raise SignatureVerificationError(self.msg("signature_mismatch"))
+
+        return SignStatusResult(
+            signature=signature,
+            signature_algorithm=signature_algorithm,
+            certificate=certificate,
+        )
+
+    # ============
+    # Internals
+    # ============
```

Comment:
If the following methods are internals why aren't they all "_method_names"

Reply:
It is legacy, but there are plans to improve that. E.g. the Mobileid and Smartid services should inherit from a common base class. So I just kept the common code close to the original, to make this merger easier.